From 15495f7777b60384bfe01baa413b23a1c5cee4e7 Mon Sep 17 00:00:00 2001
From: Lenka Doudova <lryznaro@redhat.com>
Date: Mon, 7 Jul 2025 11:28:05 +0200
Subject: [PATCH] Issue 6753 - Port and fix ticket 47823 tests

Fix and port 47823 tests into
dirsrvtests/suites/plugins/attruniq_test.py

Relates: #6753

Author: Lenka Doudova

Reviewer: Simon Pichugin
---
 .../tests/suites/plugins/attruniq_test.py     | 714 ++++++++++++-
 dirsrvtests/tests/tickets/ticket47823_test.py | 965 ------------------
 2 files changed, 694 insertions(+), 985 deletions(-)
 delete mode 100644 dirsrvtests/tests/tickets/ticket47823_test.py

diff --git a/dirsrvtests/tests/suites/plugins/attruniq_test.py b/dirsrvtests/tests/suites/plugins/attruniq_test.py
index 046952df3..ebf075ea4 100644
--- a/dirsrvtests/tests/suites/plugins/attruniq_test.py
+++ b/dirsrvtests/tests/suites/plugins/attruniq_test.py
@@ -6,17 +6,24 @@
 # See LICENSE for details.
 # --- END COPYRIGHT BLOCK ---
 #
+import subprocess
+import shutil
+import re
 import pytest
 import ldap
 import logging
+from time import sleep
 from lib389.plugins import AttributeUniquenessPlugin
 from lib389.idm.nscontainer import nsContainers
 from lib389.idm.user import UserAccounts
 from lib389.idm.group import Groups
 from lib389._constants import DEFAULT_SUFFIX
 from lib389.topologies import topology_st
+from lib389.utils import ds_is_older
 
-pytestmark = pytest.mark.tier1
+
+pytestmark = [pytest.mark.tier1,
+              pytest.mark.skipif(ds_is_older('1.3.3'), reason="Not implemented")]
 
 logging.getLogger(__name__).setLevel(logging.DEBUG)
 log = logging.getLogger(__name__)
@@ -36,8 +43,185 @@ USER_2_CN = "test_2"
 USER_3_CN = "test_3"
 USER_4_CN = "test_4"
 
+PROVISIONING_CN = "provisioning"
+PROVISIONING_DN = f"cn={PROVISIONING_CN},{DEFAULT_SUFFIX}"
+
+ACTIVE_CN = "accounts"
+STAGE_CN = "staged users"
+DELETE_CN = "deleted users"
+ACTIVE_DN = f"cn={ACTIVE_CN},{DEFAULT_SUFFIX}"
+STAGE_DN = f"cn={STAGE_CN},{PROVISIONING_DN}"
+
+ACTIVE_USER_1_CN = "active_1"
+ACTIVE_USER_2_CN = "active_2"
+
+STAGE_USER_1_CN = 'stage_1'
+STAGE_USER_2_CN = 'stage_2'
+
+
+@pytest.fixture(scope='function')
+def containers(topology_st, request):
+    """Create containers for the tests"""
+    log.info('Create containers for the tests')
+    containers = nsContainers(topology_st.standalone, DEFAULT_SUFFIX)
+    cont_provisioning = containers.create(properties={'cn': PROVISIONING_CN})
+    cont_active = containers.create(properties={'cn': ACTIVE_CN})
+
+    containers_provisioning = nsContainers(topology_st.standalone, PROVISIONING_DN)
+    cont_stage = containers_provisioning.create(properties={'cn': STAGE_CN})
+    cont_delete = containers_provisioning.create(properties={'cn': DELETE_CN})
+
+    def fin():
+        log.info('Delete containers')
+        cont_stage.delete()
+        cont_delete.delete()
+        cont_provisioning.delete()
+        cont_active.delete()
+
+    request.addfinalizer(fin)
+
+
+@pytest.fixture(scope='function')
+def attruniq(topology_st, request):
+    log.info('Setup attribute uniqueness plugin')
+    attruniq = AttributeUniquenessPlugin(topology_st.standalone, dn="cn=attruniq,cn=plugins,cn=config")
+    attruniq.create(properties={'cn': 'attruniq'})
+    attruniq.add_unique_attribute('cn')
+    topology_st.standalone.restart()
+
+    def fin():
+        if attruniq.exists():
+            attruniq.disable()
+            attruniq.delete()
+
+    request.addfinalizer(fin)
+
+    return attruniq
+
+
+@pytest.fixture(scope='function')
+def active_user_1(topology_st, request):
+    log.info('Create active user 1')
+    active_users = UserAccounts(topology_st.standalone, DEFAULT_SUFFIX,
+                                rdn='cn={}'.format(ACTIVE_CN))
+    if active_users.exists(ACTIVE_USER_1_CN):
+        active_users.get(ACTIVE_USER_1_CN).delete()
+
+    active_user_1 = active_users.create(properties={'cn': ACTIVE_USER_1_CN,
+                                               'uid': ACTIVE_USER_1_CN,
+                                               'sn': ACTIVE_USER_1_CN,
+                                               'uidNumber': '1',
+                                               'gidNumber': '11',
+                                               'homeDirectory': f'/home/{ACTIVE_USER_1_CN}'})
+
+    def fin():
+        if active_user_1.exists():
+            log.info('Delete active user 1')
+            active_user_1.delete()
+
+    request.addfinalizer(fin)
+
+    return active_user_1
+
+
+@pytest.fixture(scope='function')
+def active_user_2(topology_st, request):
+    log.info('Create active user 2')
+    active_users = UserAccounts(topology_st.standalone, DEFAULT_SUFFIX,
+                                rdn='cn={}'.format(ACTIVE_CN))
+    if active_users.exists(ACTIVE_USER_2_CN):
+        active_users.get(ACTIVE_USER_2_CN).delete()
+
+    active_user_2 = active_users.create(properties={'cn': ACTIVE_USER_2_CN,
+                                               'uid': ACTIVE_USER_2_CN,
+                                               'sn': ACTIVE_USER_2_CN,
+                                               'uidNumber': '2',
+                                               'gidNumber': '22',
+                                               'homeDirectory': f'/home/{ACTIVE_USER_2_CN}'})
+
+    def fin():
+        if active_user_2.exists():
+            log.info('Delete active user 2')
+            active_user_2.delete()
+
+    request.addfinalizer(fin)
+
+    return active_user_2
 
-def test_modrdn_attr_uniqueness(topology_st):
+
+@pytest.fixture(scope='function')
+def stage_user_1(topology_st, request):
+    log.info('Create stage user 1')
+    stage_users = UserAccounts(topology_st.standalone, DEFAULT_SUFFIX,
+                               rdn=f'cn={STAGE_CN},cn={PROVISIONING_CN}')
+    if stage_users.exists(STAGE_USER_1_CN):
+        stage_users.get(STAGE_USER_1_CN).delete()
+
+    stage_user_1 = stage_users.create(properties={'cn': STAGE_USER_1_CN,
+                                              'uid': STAGE_USER_1_CN,
+                                              'sn': STAGE_USER_1_CN,
+                                              'uidNumber': '2',
+                                              'gidNumber': '22',
+                                              'homeDirectory': f'/home/{STAGE_USER_1_CN}'})
+
+    def fin():
+        if stage_user_1.exists():
+            log.info('Delete stage user 1')
+            stage_user_1.delete()
+
+    request.addfinalizer(fin)
+
+    return stage_user_1
+
+
+@pytest.fixture(scope='function')
+def stage_user_2(topology_st, request):
+    log.info('Create stage user 1')
+    stage_users = UserAccounts(topology_st.standalone, DEFAULT_SUFFIX,
+                               rdn=f'cn={STAGE_CN},cn={PROVISIONING_CN}')
+    if stage_users.exists(STAGE_USER_2_CN):
+        stage_users.get(STAGE_USER_2_CN).delete()
+
+    stage_user_2 = stage_users.create(properties={'cn': STAGE_USER_2_CN,
+                                              'uid': STAGE_USER_2_CN,
+                                              'sn': STAGE_USER_2_CN,
+                                              'uidNumber': '2',
+                                              'gidNumber': '22',
+                                              'homeDirectory': f'/home/{STAGE_USER_2_CN}'})
+
+    def fin():
+        if stage_user_2.exists():
+            log.info('Delete stage user 2')
+            stage_user_2.delete()
+
+    request.addfinalizer(fin)
+
+    return stage_user_2
+
+
+def create_active_user(topology_st, uid=ACTIVE_USER_2_CN, cn=ACTIVE_USER_2_CN):
+    active_users = UserAccounts(topology_st.standalone, DEFAULT_SUFFIX,
+                                rdn=f'cn={ACTIVE_CN}')
+    return active_users.create(properties={'cn': cn,
+                                        'uid': uid,
+                                        'sn': uid,
+                                        'uidNumber': '2',
+                                        'gidNumber': '22',
+                                        'homeDirectory': f'/home/{uid}'})
+
+
+def create_stage_user(topology_st, uid=STAGE_USER_2_CN, cn=STAGE_USER_2_CN):
+    stage_users = UserAccounts(topology_st.standalone, DEFAULT_SUFFIX,
+                                rdn=f'cn={STAGE_CN},cn={PROVISIONING_CN}')
+    return stage_users.create(properties={'cn': cn,
+                                        'uid': uid,
+                                        'sn': uid,
+                                        'uidNumber': '3',
+                                        'gidNumber': '33',
+                                        'homeDirectory': f'/home/{uid}'})
+
+
+def test_modrdn_attr_uniqueness(topology_st, attruniq):
     """Test that we can not add two entries that have the same attr value that is
     defined by the plugin
 
@@ -67,9 +251,6 @@ def test_modrdn_attr_uniqueness(topology_st):
     group1 = groups.create(properties={'cn': 'group1'})
     group2 = groups.create(properties={'cn': 'group2'})
 
-    attruniq = AttributeUniquenessPlugin(topology_st.standalone, dn="cn=attruniq,cn=plugins,cn=config")
-    log.debug(f'Setup PLUGIN_ATTR_UNIQUENESS plugin for {MAIL_ATTR_VALUE} attribute for the group2')
-    attruniq.create(properties={'cn': 'attruniq'})
     attruniq.add_unique_attribute('mail')
     attruniq.add_unique_subtree(group2.dn)
     attruniq.enable_all_subtrees()
@@ -97,14 +278,11 @@ def test_modrdn_attr_uniqueness(topology_st):
     log.debug('Move user2 to group1')
     user2.rename(f'uid={user2.rdn}', group1.dn)
 
-    # Cleanup for next test
     user1.delete()
     user2.delete()
-    attruniq.disable()
-    attruniq.delete()
 
 
-def test_multiple_attr_uniqueness(topology_st):
+def test_multiple_attr_uniqueness(topology_st, attruniq):
     """ Test that attribute uniqueness works properly with multiple attributes
 
     :id: c49aa5c1-7e65-45fd-b064-55e0b815e9bc
@@ -124,11 +302,8 @@ def test_multiple_attr_uniqueness(topology_st):
         5. Should raise CONSTRAINT_VIOLATION
         6. Should raise CONSTRAINT_VIOLATION
     """
-    attruniq = AttributeUniquenessPlugin(topology_st.standalone, dn="cn=attruniq,cn=plugins,cn=config")
 
     try:
-        log.debug(f'Setup PLUGIN_ATTR_UNIQUENESS plugin for {MAIL_ATTR_VALUE} attribute for the group2')
-        attruniq.create(properties={'cn': 'attruniq'})
         attruniq.add_unique_attribute('mail')
         attruniq.add_unique_attribute('mailAlternateAddress')
         attruniq.add_unique_subtree(DEFAULT_SUFFIX)
@@ -163,14 +338,11 @@ def test_multiple_attr_uniqueness(topology_st):
     with pytest.raises(ldap.CONSTRAINT_VIOLATION):
         testuser2.add('mailAlternateAddress', MAIL_ATTR_VALUE)
 
-    # Cleanup
     testuser1.delete()
     testuser2.delete()
-    attruniq.disable()
-    attruniq.delete()
 
 
-def test_exclude_subtrees(topology_st):
+def test_exclude_subtrees(topology_st, attruniq):
     """ Test attribute uniqueness with exclude scope
 
     :id: 43d29a60-40e1-4ebd-b897-6ef9f20e9f27
@@ -212,9 +384,6 @@ def test_exclude_subtrees(topology_st):
         16. Success
         17. Success
     """
-    log.info('Setup attribute uniqueness plugin')
-    attruniq = AttributeUniquenessPlugin(topology_st.standalone, dn="cn=attruniq,cn=plugins,cn=config")
-    attruniq.create(properties={'cn': 'attruniq'})
     attruniq.add_unique_attribute('telephonenumber')
     attruniq.add_unique_subtree(DEFAULT_SUFFIX)
     attruniq.enable_all_subtrees()
@@ -340,7 +509,6 @@ def test_matchingrule_attr(topology_st):
         3. Add second user with same lowercase CN which should be rejected
         4. Add second user with same CN value but with mixed case
         5. Modify second user replacing CN value to lc which should be rejected
-
     :expectedresults:
         1. Success
         2. Success
@@ -386,3 +554,509 @@ def test_matchingrule_attr(topology_st):
     log.info('Mod entry with exact case CN value which should be rejected')
     with pytest.raises(ldap.CONSTRAINT_VIOLATION):
         user.replace('cn', 'common_name')
+
+    attruniq.disable()
+    attruniq.delete()
+    inst.restart()
+
+
+def test_one_container_add(topology_st, attruniq, containers, active_user_1):
+    """Test ADD operations with attribute uniqueness in a single container.
+
+    This test verifies that the attribute uniqueness plugin correctly prevents
+    the addition of entries with duplicate attribute values within a single
+    container when the plugin is enabled.
+
+    :id: 05960fb7-3ef8-4b53-b616-126f7ac737c4
+    :setup: Standalone instance with containers and one active user
+    :steps:
+        1. Verify user with duplicate 'cn' attribute can be added
+           without enforced uniqueness
+        2. Remove the duplicate user
+        3. Configure and enable attribute uniqueness plugin for 'cn' in active container
+        4. Try to add user with duplicate 'cn' attribute (should fail)
+    :expectedresults:
+        1. Success
+        2. Success
+        3. Success
+        4. Duplicate entry creation fails with CONSTRAINT_VIOLATION
+    """
+
+    log.info('Verify that there is no uniqueness rule enforced')
+    active_users = UserAccounts(topology_st.standalone, DEFAULT_SUFFIX,
+                                rdn='cn={}'.format(ACTIVE_CN))
+    active_2 = active_users.create(properties={'cn': [ACTIVE_USER_1_CN, ACTIVE_USER_2_CN],
+                                               'uid': ACTIVE_USER_2_CN,
+                                               'sn': ACTIVE_USER_2_CN,
+                                               'uidNumber': '2',
+                                               'gidNumber': '22',
+                                               'homeDirectory': f'/home/{ACTIVE_USER_2_CN}'})
+
+    log.info('Remove the second user')
+    active_2.delete()
+
+    log.info('Setup attribute uniqueness plugin for "cn" attribute')
+    attruniq.add_unique_subtree(ACTIVE_DN)
+    attruniq.enable()
+    topology_st.standalone.restart()
+
+    log.info('Verify that uniqueness rule is enforced')
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        active_users.create(properties={'cn': [ACTIVE_USER_1_CN, ACTIVE_USER_2_CN],
+                                                   'uid': ACTIVE_USER_2_CN,
+                                                   'sn': ACTIVE_USER_2_CN,
+                                                   'uidNumber': '2',
+                                                   'gidNumber': '22',
+                                                   'homeDirectory': f'/home/{ACTIVE_USER_2_CN}'})
+
+
+def test_one_container_mod(topology_st, attruniq, containers,
+                           active_user_1, active_user_2):
+    """Test MOD operations with attribute uniqueness in a single container.
+
+    This test verifies that the attribute uniqueness plugin correctly prevents
+    modification of entries that would create duplicate attribute values within
+    a single container.
+
+    :id: c566f279-6073-4fb7-8956-f30d1bdda8e2
+    :setup: Standalone instance with containers and two active users
+    :steps:
+        1. Configure and enable attribute uniqueness plugin for "cn" attribute
+        2. Try to modify user2's "cn" to match user1's "cn" (should fail)
+    :expectedresults:
+        1. Plugin configured and enabled successfully
+        2. Modification fails with CONSTRAINT_VIOLATION
+    """
+
+    log.info('Setup attribute uniqueness plugin for "cn" attribute')
+    attruniq.add_unique_subtree(ACTIVE_DN)
+    attruniq.enable()
+    topology_st.standalone.restart()
+
+    log.info('Verify that uniqueness rule is enforced')
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        active_user_2.replace('cn', ACTIVE_USER_1_CN)
+
+
+@pytest.mark.xfail(reason="DS6846")
+def test_one_container_modrdn(topology_st, attruniq, containers,
+                              active_user_1, active_user_2):
+    """Test MODRDN operations with attribute uniqueness in a single container.
+
+    This test verifies that the attribute uniqueness plugin correctly prevents
+    ModRDN operations that would create duplicate attribute values within
+    a single container.
+
+    :id: cabfec46-bea7-4175-913e-2f7ab84a2919
+    :setup: Standalone instance with containers and two active users
+    :steps:
+        1. Configure and enable attribute uniqueness plugin for "cn" attribute
+        2. Try to rename user2 to have the same "cn" as user1 (should fail)
+    :expectedresults:
+        1. Plugin configured and enabled successfully
+        2. ModRDN operation fails with CONSTRAINT_VIOLATION
+    """
+
+    log.info('Setup attribute uniqueness plugin for "cn" attribute')
+    attruniq.add_unique_subtree(ACTIVE_DN)
+    attruniq.enable()
+    topology_st.standalone.restart()
+
+    log.info('Verify that uniqueness rule is enforced')
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        active_user_2.rename(f'cn={ACTIVE_USER_1_CN}', deloldrdn=False)
+
+
+def test_multiple_containers_add(topology_st, attruniq, containers,
+                                  active_user_1, stage_user_1):
+    """Test ADD operations with attribute uniqueness across multiple containers.
+
+    This test verifies that when multiple containers are configured for
+    attribute uniqueness, the plugin enforces uniqueness within each container
+    separately, but allows duplicate values across different containers.
+
+    :id: 400cdf54-499a-4c23-91f5-835f44835e3e
+    :setup: Standalone instance with containers and users in different containers
+    :steps:
+        1. Configure plugin for 'cn' attribute in both active and stage containers
+        2. Try to add user with duplicate 'cn' in active container (should fail)
+        3. Try to add user with duplicate 'cn' in stage container (should fail)
+        4. Try to add user in active container with 'cn' matching stage user (should succeed)
+        5. Try to add user in stage container with 'cn' matching active user (should succeed)
+    :expectedresults:
+        1. Plugin configured successfully
+        2. Duplicate in active container fails with CONSTRAINT_VIOLATION
+        3. Duplicate in stage container fails with CONSTRAINT_VIOLATION
+        4. Cross-container duplicate in active succeeds
+        5. Cross-container duplicate in stage succeeds
+    """
+
+    log.info('Setup attribute uniqueness plugin for "cn" attribute')
+    attruniq.add_unique_subtree(ACTIVE_DN)
+    attruniq.add_unique_subtree(STAGE_DN)
+    attruniq.enable()
+    topology_st.standalone.restart()
+
+    log.info("Verify both subtrees apply attruniqueness separately")
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        create_active_user(topology_st, uid=ACTIVE_USER_2_CN, cn=[ACTIVE_USER_1_CN, ACTIVE_USER_2_CN])
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        create_stage_user(topology_st, uid=STAGE_USER_2_CN, cn=[STAGE_USER_1_CN, STAGE_USER_2_CN])
+
+    log.info("Verify that the subtrees do not enforce uniqueness across each other")
+    try:
+        active_user_2 = create_active_user(topology_st, uid=ACTIVE_USER_2_CN,
+                                           cn=[STAGE_USER_1_CN, ACTIVE_USER_2_CN])
+        stage_user_2 = create_stage_user(topology_st, uid=STAGE_USER_2_CN,
+                                         cn=[ACTIVE_USER_1_CN, STAGE_USER_2_CN])
+    except:
+        assert False, "Failed to create users necessary for the test"
+    finally:
+        active_user_2.delete()
+        stage_user_2.delete()
+
+
+def test_multiple_containers_add_across_subtrees(topology_st, attruniq, containers,
+                           active_user_1, stage_user_1):
+    """Test ADD operations with attribute uniqueness across multiple containers with cross-subtree enforcement.
+
+    This test verifies that when the 'across-all-subtrees' option is enabled,
+    the attribute uniqueness plugin enforces uniqueness across all configured
+    containers, preventing duplicate values anywhere in the configured scope.
+
+    :id: 7b58c46a-9ef1-4bbc-929d-578dbd3f4910
+    :setup: Standalone instance with containers and users in different containers
+    :steps:
+        1. Configure plugin for 'cn' attribute in both containers with across-subtrees enabled
+        2. Try to add user with duplicate 'cn' in active container (should fail)
+        3. Try to add user with duplicate 'cn' in stage container (should fail)
+        4. Try to add user in stage container with 'cn' matching active user (should fail)
+        5. Try to add user in active container with 'cn' matching stage user (should fail)
+    :expectedresults:
+        1. Plugin configured successfully with across-subtrees enabled
+        2. Duplicate in active container fails with CONSTRAINT_VIOLATION
+        3. Duplicate in stage container fails with CONSTRAINT_VIOLATION
+        4. Cross-container duplicate in stage fails with CONSTRAINT_VIOLATION
+        5. Cross-container duplicate in active fails with CONSTRAINT_VIOLATION
+    """
+
+    log.info('Setup attribute uniqueness plugin for "cn" attribute')
+    attruniq.add_unique_subtree(ACTIVE_DN)
+    attruniq.add_unique_subtree(STAGE_DN)
+    attruniq.enable_all_subtrees()
+    attruniq.enable()
+    topology_st.standalone.restart()
+
+    # Inserting sleep to prevent repeatedly occuring 'Server is busy' errors
+    sleep(5)
+
+    log.info("Verify both subtrees apply attruniqueness separately")
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        create_active_user(topology_st, uid=ACTIVE_USER_2_CN, cn=[ACTIVE_USER_1_CN, ACTIVE_USER_2_CN])
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        create_stage_user(topology_st, uid=STAGE_USER_2_CN, cn=[STAGE_USER_1_CN, STAGE_USER_2_CN])
+
+    log.info('Verify that uniqueness rule is enforced')
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        create_stage_user(topology_st, uid=STAGE_USER_2_CN, cn=[ACTIVE_USER_1_CN, STAGE_USER_2_CN])
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        create_active_user(topology_st, uid=ACTIVE_USER_2_CN, cn=[STAGE_USER_1_CN, ACTIVE_USER_2_CN])
+
+
+def test_multiple_containers_mod(topology_st, attruniq, containers,
+                           active_user_1, active_user_2, stage_user_1, stage_user_2):
+    """Test MOD operations with attribute uniqueness across multiple containers.
+
+    This test verifies that when multiple containers are configured for
+    attribute uniqueness without cross-subtree enforcement, modifications
+    are constrained within each container but allowed across containers.
+
+    :id: 29ad0e41-bb6f-4ec0-bf79-0d6a4402a5db
+    :setup: Standalone instance with containers and users in different containers
+    :steps:
+        1. Configure plugin for 'cn' attribute in both containers
+        2. Try to modify user in active container to match another active user (should fail)
+        3. Try to modify user in stage container to match another stage user (should fail)
+        4. Try to modify user in active container to match stage user (should succeed)
+        5. Try to modify user in stage container to match active user (should succeed)
+    :expectedresults:
+        1. Plugin configured successfully
+        2. Intra-container modification fails with CONSTRAINT_VIOLATION
+        3. Intra-container modification fails with CONSTRAINT_VIOLATION
+        4. Cross-container modification succeeds
+        5. Cross-container modification succeeds
+    """
+
+    log.info('Setup attribute uniqueness plugin for "cn" attribute')
+    attruniq.add_unique_subtree(ACTIVE_DN)
+    attruniq.add_unique_subtree(STAGE_DN)
+    attruniq.enable()
+    topology_st.standalone.restart()
+
+    log.info('Verify that uniqueness is enforced in individual subtrees')
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        active_user_1.replace('cn', ACTIVE_USER_2_CN)
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        stage_user_1.replace('cn', STAGE_USER_2_CN)
+
+    log.info('Verify that uniqueness rule is not enforced across subtrees')
+    active_user_1.replace('cn', STAGE_USER_1_CN)
+    stage_user_1.replace('cn', ACTIVE_USER_1_CN)
+
+
+def test_multiple_containers_mod_across_subtrees(topology_st, attruniq, containers,
+                           active_user_1, active_user_2, stage_user_1, stage_user_2):
+    """Test MOD operations with attribute uniqueness across multiple containers with cross-subtree enforcement.
+
+    This test verifies that when the 'across-all-subtrees' option is enabled,
+    modification operations are constrained across all configured containers,
+    preventing duplicate values anywhere in the configured scope.
+
+    :id: f5272114-8869-4c61-96e2-0241500c47c2
+    :setup: Standalone instance with containers and users in different containers
+    :steps:
+        1. Configure plugin for 'cn' attribute in both containers with across-subtrees enabled
+        2. Try to modify user in active container to match another active user (should fail)
+        3. Try to modify user in stage container to match another stage user (should fail)
+        4. Try to modify user in active container to match stage user (should fail)
+        5. Try to modify user in stage container to match active user (should fail)
+    :expectedresults:
+        1. Plugin configured successfully with across-subtrees enabled
+        2. Intra-container modification fails with CONSTRAINT_VIOLATION
+        3. Intra-container modification fails with CONSTRAINT_VIOLATION
+        4. Cross-container modification fails with CONSTRAINT_VIOLATION
+        5. Cross-container modification fails with CONSTRAINT_VIOLATION
+    """
+
+    log.info('Setup attribute uniqueness plugin for "cn" attribute')
+    attruniq.add_unique_subtree(ACTIVE_DN)
+    attruniq.add_unique_subtree(STAGE_DN)
+    attruniq.enable_all_subtrees()
+    attruniq.enable()
+    topology_st.standalone.restart()
+
+    log.info('Verify that uniqueness is enforced in individual subtrees')
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        active_user_1.replace('cn', ACTIVE_USER_2_CN)
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        stage_user_1.replace('cn', STAGE_USER_2_CN)
+
+    log.info('Verify that uniqueness rule is enforced across subtrees')
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        active_user_1.replace('cn', STAGE_USER_1_CN)
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        stage_user_1.replace('cn', ACTIVE_USER_1_CN)
+
+
+@pytest.mark.xfail(reason="DS6846")
+def test_multiple_containers_modrdn(topology_st, attruniq, containers,
+                           active_user_1, active_user_2, stage_user_1, stage_user_2):
+    """Test MODRDN operations with attribute uniqueness across multiple containers.
+
+    This test verifies that when multiple containers are configured for
+    attribute uniqueness without cross-subtree enforcement, ModRDN operations
+    are constrained within each container but allowed across containers.
+
+    :id: 4b835bfe-d116-418a-beb9-943e23b7cd2a
+    :setup: Standalone instance with containers and users in different containers
+    :steps:
+        1. Configure plugin for 'cn' attribute in both containers
+        2. Try to rename user in active container to match another active user (should fail)
+        3. Try to rename user in stage container to match another stage user (should fail)
+        4. Try to rename user in active container to match stage user (should succeed)
+        5. Try to rename user in stage container to match active user (should succeed)
+    :expectedresults:
+        1. Plugin configured successfully
+        2. Intra-container ModRDN fails with CONSTRAINT_VIOLATION
+        3. Intra-container ModRDN fails with CONSTRAINT_VIOLATION
+        4. Cross-container ModRDN succeeds
+        5. Cross-container ModRDN succeeds
+    """
+
+    log.info('Setup attribute uniqueness plugin for "cn" attribute')
+    attruniq.add_unique_subtree(ACTIVE_DN)
+    attruniq.add_unique_subtree(STAGE_DN)
+    attruniq.enable()
+    topology_st.standalone.restart()
+
+    log.info('Verify that uniqueness is enforced in individual subtrees')
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        active_user_1.rename(f'cn={ACTIVE_USER_2_CN}', deloldrdn=False)
+    with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+        stage_user_1.rename(f'cn={STAGE_USER_2_CN}', deloldrdn=False)
+
+    log.info('Verify that uniqueness rule is not enforced across subtrees')
+    active_user_2.rename(f'cn={STAGE_USER_1_CN}', deloldrdn=False)
+    stage_user_2.rename(f'cn={ACTIVE_USER_1_CN}', deloldrdn=False)
+
+
+@pytest.mark.xfail(reason="DS6846")
+def test_multiple_containers_modrdn_across_subtrees(topology_st, attruniq, containers,
+                                        active_user_1, active_user_2, stage_user_1, stage_user_2):
+        """Test MODRDN operations with attribute uniqueness across multiple containers with cross-subtree enforcement.
+
+        This test verifies that when the 'across-all-subtrees' option is enabled,
+        ModRDN operations are constrained across all configured containers,
+        preventing duplicate values anywhere in the configured scope.
+
+        :id: 8d2c4af9-1b45-4c72-9523-7f8b39c74e21
+        :setup: Standalone instance with containers and users in different containers
+        :steps:
+            1. Configure plugin for 'cn' attribute in both containers with across-subtrees enabled
+            2. Try to rename user in active container to match another active user (should fail)
+            3. Try to rename user in stage container to match another stage user (should fail)
+            4. Try to rename user in active container to match stage user (should fail)
+            5. Try to rename user in stage container to match active user (should fail)
+        :expectedresults:
+            1. Plugin configured successfully with across-subtrees enabled
+            2. Intra-container ModRDN fails with CONSTRAINT_VIOLATION
+            3. Intra-container ModRDN fails with CONSTRAINT_VIOLATION
+            4. Cross-container ModRDN fails with CONSTRAINT_VIOLATION
+            5. Cross-container ModRDN fails with CONSTRAINT_VIOLATION
+        """
+
+        log.info('Setup attribute uniqueness plugin for "cn" attribute')
+        attruniq.add_unique_subtree(ACTIVE_DN)
+        attruniq.add_unique_subtree(STAGE_DN)
+        attruniq.enable_all_subtrees()
+        attruniq.enable()
+        topology_st.standalone.restart()
+
+        log.info('Verify that uniqueness is enforced in individual subtrees')
+        with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+            active_user_1.rename(f'cn={ACTIVE_USER_2_CN}', deloldrdn=False)
+        with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+            stage_user_1.rename(f'cn={STAGE_USER_2_CN}', deloldrdn=False)
+
+        log.info('Verify that uniqueness rule is enforced across subtrees')
+        with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+            active_user_2.rename(f'cn={STAGE_USER_1_CN}', deloldrdn=False)
+        with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+            stage_user_2.rename(f'cn={ACTIVE_USER_1_CN}', deloldrdn=False)
+
+
+def _config_file(topology_st, action='save'):
+    dse_ldif = topology_st.standalone.confdir + '/dse.ldif'
+    sav_file = topology_st.standalone.confdir + '/dse.ldif.ticket47823'
+    if action == 'save':
+        shutil.copy(dse_ldif, sav_file)
+    else:
+        shutil.copy(sav_file, dse_ldif)
+
+
+def _pattern_errorlog(file, log_pattern):
+    try:
+        _pattern_errorlog.last_pos += 1
+    except AttributeError:
+        _pattern_errorlog.last_pos = 0
+
+    found = None
+    log.debug("_pattern_errorlog: start at offset %d" % _pattern_errorlog.last_pos)
+    file.seek(_pattern_errorlog.last_pos)
+
+    while True:
+        line = file.readline()
+        log.debug("_pattern_errorlog: [%d] %s" % (file.tell(), line))
+        found = log_pattern.search(line)
+        if ((line == '') or (found)):
+            break
+
+    log.debug("_pattern_errorlog: end at offset %d" % file.tell())
+    _pattern_errorlog.last_pos = file.tell()
+    return found
+
+
+def test_invalid_config_missing_attr_name(topology_st):
+    """Test that invalid plugin configuration is properly detected when attribute name is missing.
+
+    This test verifies that the attribute uniqueness plugin correctly fails
+    to start when the required attribute name configuration is missing,
+    and that appropriate error messages are logged.
+
+    :id: 2ff3aa56-304e-41b0-9614-eeb481ef6c55
+    :setup: Standalone instance
+    :steps:
+        1. Save current configuration
+        2. Create plugin configuration without uniqueness attribute name
+        3. Enable the plugin
+        4. Try to restart the server (should fail)
+        5. Verify expected error message in logs
+        6. Restore configuration and verify server can restart
+    :expectedresults:
+        1. Configuration saved successfully
+        2. Invalid plugin configuration created
+        3. Plugin enabled successfully
+        4. Server restart fails with appropriate error
+        5. Error message "Attribute name not defined" found in logs
+        6. Server restarts successfully with restored configuration
+    """
+
+    _config_file(topology_st, action='save')
+
+    attruniq = AttributeUniquenessPlugin(topology_st.standalone, dn="cn=attruniq,cn=plugins,cn=config")
+    attruniq.create(properties={'cn': 'attruniq'})
+    attruniq.enable()
+
+    topology_st.standalone.errorlog_file = open(topology_st.standalone.errlog, "r")
+
+    with pytest.raises(subprocess.CalledProcessError):
+        log.info("Server is down as expected due to invalid config")
+        topology_st.standalone.restart()
+
+    log.info("Check the expected error message")
+    regex = re.compile("[A|a]ttribute name not defined")
+    res = _pattern_errorlog(topology_st.standalone.errorlog_file, regex)
+    assert res
+
+    log.info("Restore the configuration and verify the server can be restarted")
+    _config_file(topology_st, action='restore')
+    topology_st.standalone.restart()
+
+
+def test_invalid_config_invalid_subtree(topology_st):
+    """Test that invalid plugin configuration is properly detected when subtree is invalid.
+
+    This test verifies that the attribute uniqueness plugin correctly fails
+    to start when an invalid subtree DN is specified in the configuration,
+    and that appropriate error messages are logged.
+
+    :id: b7e51131-8542-4110-a90a-3e5c88f75355
+    :setup: Standalone instance
+    :steps:
+        1. Save current configuration
+        2. Create plugin configuration with invalid subtree DN
+        3. Enable the plugin
+        4. Try to restart the server (should fail)
+        5. Verify expected error message in logs
+        6. Restore configuration and verify server can restart
+    :expectedresults:
+        1. Configuration saved successfully
+        2. Invalid plugin configuration created with bad subtree
+        3. Plugin enabled successfully
+        4. Server restart fails with appropriate error
+        5. Error message "No valid subtree is defined" found in logs
+        6. Server restarts successfully with restored configuration
+    """
+
+    _config_file(topology_st, action='save')
+
+    attruniq = AttributeUniquenessPlugin(topology_st.standalone, dn="cn=attruniq,cn=plugins,cn=config")
+    attruniq.create(properties={'cn': 'attruniq'})
+    attruniq.add_unique_attribute('cn')
+    attruniq.add_unique_subtree('invalid_subtree')
+    attruniq.enable()
+
+    topology_st.standalone.errorlog_file = open(topology_st.standalone.errlog, "r")
+
+    with pytest.raises(subprocess.CalledProcessError):
+        log.info("Server is down as expected due to invalid config")
+        topology_st.standalone.restart()
+
+    log.info("Check the expected error message")
+    regex = re.compile("No valid subtree is defined")
+    res = _pattern_errorlog(topology_st.standalone.errorlog_file, regex)
+    assert res
+
+    log.info("Restore the configuration and verify the server can be restarted")
+    _config_file(topology_st, action='restore')
+    topology_st.standalone.restart()
diff --git a/dirsrvtests/tests/tickets/ticket47823_test.py b/dirsrvtests/tests/tickets/ticket47823_test.py
deleted file mode 100644
index 07a3d9062..000000000
--- a/dirsrvtests/tests/tickets/ticket47823_test.py
+++ /dev/null
@@ -1,965 +0,0 @@
-# --- BEGIN COPYRIGHT BLOCK ---
-# Copyright (C) 2016 Red Hat, Inc.
-# All rights reserved.
-#
-# License: GPL (version 3 or any later version).
-# See LICENSE for details.
-# --- END COPYRIGHT BLOCK ---
-#
-import logging
-import re
-import shutil
-import subprocess
-import time
-
-import ldap
-import pytest
-from lib389 import Entry
-from lib389._constants import *
-from lib389.topologies import topology_st
-
-log = logging.getLogger(__name__)
-from lib389.utils import *
-
-# Skip on older versions
-pytestmark = [pytest.mark.tier2,
-              pytest.mark.skipif(ds_is_older('1.3.3'), reason="Not implemented")]
-PROVISIONING_CN = "provisioning"
-PROVISIONING_DN = "cn=%s,%s" % (PROVISIONING_CN, SUFFIX)
-
-ACTIVE_CN = "accounts"
-STAGE_CN = "staged users"
-DELETE_CN = "deleted users"
-ACTIVE_DN = "cn=%s,%s" % (ACTIVE_CN, SUFFIX)
-STAGE_DN = "cn=%s,%s" % (STAGE_CN, PROVISIONING_DN)
-DELETE_DN = "cn=%s,%s" % (DELETE_CN, PROVISIONING_DN)
-
-STAGE_USER_CN = "stage guy"
-STAGE_USER_DN = "cn=%s,%s" % (STAGE_USER_CN, STAGE_DN)
-
-ACTIVE_USER_CN = "active guy"
-ACTIVE_USER_DN = "cn=%s,%s" % (ACTIVE_USER_CN, ACTIVE_DN)
-
-ACTIVE_USER_1_CN = "test_1"
-ACTIVE_USER_1_DN = "cn=%s,%s" % (ACTIVE_USER_1_CN, ACTIVE_DN)
-ACTIVE_USER_2_CN = "test_2"
-ACTIVE_USER_2_DN = "cn=%s,%s" % (ACTIVE_USER_2_CN, ACTIVE_DN)
-
-STAGE_USER_1_CN = ACTIVE_USER_1_CN
-STAGE_USER_1_DN = "cn=%s,%s" % (STAGE_USER_1_CN, STAGE_DN)
-STAGE_USER_2_CN = ACTIVE_USER_2_CN
-STAGE_USER_2_DN = "cn=%s,%s" % (STAGE_USER_2_CN, STAGE_DN)
-
-ALL_CONFIG_ATTRS = ['nsslapd-pluginarg0', 'nsslapd-pluginarg1', 'nsslapd-pluginarg2',
-                    'uniqueness-attribute-name', 'uniqueness-subtrees', 'uniqueness-across-all-subtrees']
-
-
-def _header(topology_st, label):
-    topology_st.standalone.log.info("\n\n###############################################")
-    topology_st.standalone.log.info("#######")
-    topology_st.standalone.log.info("####### %s" % label)
-    topology_st.standalone.log.info("#######")
-    topology_st.standalone.log.info("###############################################")
-
-
-def _uniqueness_config_entry(topology_st, name=None):
-    if not name:
-        return None
-
-    ent = topology_st.standalone.getEntry("cn=%s,%s" % (PLUGIN_ATTR_UNIQUENESS, DN_PLUGIN), ldap.SCOPE_BASE,
-                                          "(objectclass=nsSlapdPlugin)",
-                                          ['objectClass', 'cn', 'nsslapd-pluginPath', 'nsslapd-pluginInitfunc',
-                                           'nsslapd-pluginType', 'nsslapd-pluginEnabled',
-                                           'nsslapd-plugin-depends-on-type',
-                                           'nsslapd-pluginId', 'nsslapd-pluginVersion', 'nsslapd-pluginVendor',
-                                           'nsslapd-pluginDescription'])
-    ent.dn = "cn=%s uniqueness,%s" % (name, DN_PLUGIN)
-    return ent
-
-
-def _build_config(topology_st, attr_name='cn', subtree_1=None, subtree_2=None, type_config='old',
-                  across_subtrees=False):
-    assert topology_st
-    assert attr_name
-    assert subtree_1
-
-    if type_config == 'old':
-        # enable the 'cn' uniqueness on Active
-        config = _uniqueness_config_entry(topology_st, attr_name)
-        config.setValue('nsslapd-pluginarg0', attr_name)
-        config.setValue('nsslapd-pluginarg1', subtree_1)
-        if subtree_2:
-            config.setValue('nsslapd-pluginarg2', subtree_2)
-    else:
-        # prepare the config entry
-        config = _uniqueness_config_entry(topology_st, attr_name)
-        config.setValue('uniqueness-attribute-name', attr_name)
-        config.setValue('uniqueness-subtrees', subtree_1)
-        if subtree_2:
-            config.setValue('uniqueness-subtrees', subtree_2)
-        if across_subtrees:
-            config.setValue('uniqueness-across-all-subtrees', 'on')
-    return config
-
-
-def _active_container_invalid_cfg_add(topology_st):
-    '''
-    Check uniqueness is not enforced with ADD (invalid config)
-    '''
-    topology_st.standalone.add_s(Entry((ACTIVE_USER_1_DN, {
-        'objectclass': "top person".split(),
-        'sn': ACTIVE_USER_1_CN,
-        'cn': ACTIVE_USER_1_CN})))
-
-    topology_st.standalone.add_s(Entry((ACTIVE_USER_2_DN, {
-        'objectclass': "top person".split(),
-        'sn': ACTIVE_USER_2_CN,
-        'cn': [ACTIVE_USER_1_CN, ACTIVE_USER_2_CN]})))
-
-    topology_st.standalone.delete_s(ACTIVE_USER_1_DN)
-    topology_st.standalone.delete_s(ACTIVE_USER_2_DN)
-
-
-def _active_container_add(topology_st, type_config='old'):
-    '''
-    Check uniqueness in a single container (Active)
-    Add an entry with a given 'cn', then check we can not add an entry with the same 'cn' value
-
-    '''
-    config = _build_config(topology_st, attr_name='cn', subtree_1=ACTIVE_DN, subtree_2=None, type_config=type_config,
-                           across_subtrees=False)
-
-    # remove the 'cn' uniqueness entry
-    try:
-        topology_st.standalone.delete_s(config.dn)
-
-    except ldap.NO_SUCH_OBJECT:
-        pass
-    topology_st.standalone.restart(timeout=120)
-
-    topology_st.standalone.log.info('Uniqueness not enforced: create the entries')
-
-    topology_st.standalone.add_s(Entry((ACTIVE_USER_1_DN, {
-        'objectclass': "top person".split(),
-        'sn': ACTIVE_USER_1_CN,
-        'cn': ACTIVE_USER_1_CN})))
-
-    topology_st.standalone.add_s(Entry((ACTIVE_USER_2_DN, {
-        'objectclass': "top person".split(),
-        'sn': ACTIVE_USER_2_CN,
-        'cn': [ACTIVE_USER_1_CN, ACTIVE_USER_2_CN]})))
-
-    topology_st.standalone.delete_s(ACTIVE_USER_1_DN)
-    topology_st.standalone.delete_s(ACTIVE_USER_2_DN)
-
-    topology_st.standalone.log.info('Uniqueness enforced: checks second entry is rejected')
-
-    # enable the 'cn' uniqueness on Active
-    topology_st.standalone.add_s(config)
-    topology_st.standalone.restart(timeout=120)
-    topology_st.standalone.add_s(Entry((ACTIVE_USER_1_DN, {
-        'objectclass': "top person".split(),
-        'sn': ACTIVE_USER_1_CN,
-        'cn': ACTIVE_USER_1_CN})))
-
-    try:
-        topology_st.standalone.add_s(Entry((ACTIVE_USER_2_DN, {
-            'objectclass': "top person".split(),
-            'sn': ACTIVE_USER_2_CN,
-            'cn': [ACTIVE_USER_1_CN, ACTIVE_USER_2_CN]})))
-    except ldap.CONSTRAINT_VIOLATION:
-        # yes it is expected
-        pass
-
-    # cleanup the stuff now
-    topology_st.standalone.delete_s(config.dn)
-    topology_st.standalone.delete_s(ACTIVE_USER_1_DN)
-
-
-def _active_container_mod(topology_st, type_config='old'):
-    '''
-    Check uniqueness in a single container (active)
-    Add and entry with a given 'cn', then check we can not modify an entry with the same 'cn' value
-
-    '''
-
-    config = _build_config(topology_st, attr_name='cn', subtree_1=ACTIVE_DN, subtree_2=None, type_config=type_config,
-                           across_subtrees=False)
-
-    # enable the 'cn' uniqueness on Active
-    topology_st.standalone.add_s(config)
-    topology_st.standalone.restart(timeout=120)
-
-    topology_st.standalone.log.info('Uniqueness enforced: checks MOD ADD entry is rejected')
-    topology_st.standalone.add_s(Entry((ACTIVE_USER_1_DN, {
-        'objectclass': "top person".split(),
-        'sn': ACTIVE_USER_1_CN,
-        'cn': ACTIVE_USER_1_CN})))
-
-    topology_st.standalone.add_s(Entry((ACTIVE_USER_2_DN, {
-        'objectclass': "top person".split(),
-        'sn': ACTIVE_USER_2_CN,
-        'cn': ACTIVE_USER_2_CN})))
-
-    try:
-        topology_st.standalone.modify_s(ACTIVE_USER_2_DN, [(ldap.MOD_ADD, 'cn', ensure_bytes(ACTIVE_USER_1_CN))])
-    except ldap.CONSTRAINT_VIOLATION:
-        # yes it is expected
-        pass
-
-    topology_st.standalone.log.info('Uniqueness enforced: checks MOD REPLACE entry is rejected')
-    try:
-        topology_st.standalone.modify_s(ACTIVE_USER_2_DN,
-                                        [(ldap.MOD_REPLACE, 'cn', [ensure_bytes(ACTIVE_USER_1_CN), ensure_bytes(ACTIVE_USER_2_CN)])])
-    except ldap.CONSTRAINT_VIOLATION:
-        # yes it is expected
-        pass
-
-    # cleanup the stuff now
-    topology_st.standalone.delete_s(config.dn)
-    topology_st.standalone.delete_s(ACTIVE_USER_1_DN)
-    topology_st.standalone.delete_s(ACTIVE_USER_2_DN)
-
-
-def _active_container_modrdn(topology_st, type_config='old'):
-    '''
-    Check uniqueness in a single container
-    Add and entry with a given 'cn', then check we can not modrdn an entry with the same 'cn' value
-
-    '''
-    config = _build_config(topology_st, attr_name='cn', subtree_1=ACTIVE_DN, subtree_2=None, type_config=type_config,
-                           across_subtrees=False)
-
-    # enable the 'cn' uniqueness on Active
-    topology_st.standalone.add_s(config)
-    topology_st.standalone.restart(timeout=120)
-
-    topology_st.standalone.log.info('Uniqueness enforced: checks MODRDN entry is rejected')
-
-    topology_st.standalone.add_s(Entry((ACTIVE_USER_1_DN, {
-        'objectclass': "top person".split(),
-        'sn': ACTIVE_USER_1_CN,
-        'cn': [ACTIVE_USER_1_CN, 'dummy']})))
-
-    topology_st.standalone.add_s(Entry((ACTIVE_USER_2_DN, {
-        'objectclass': "top person".split(),
-        'sn': ACTIVE_USER_2_CN,
-        'cn': ACTIVE_USER_2_CN})))
-
-    try:
-        topology_st.standalone.rename_s(ACTIVE_USER_2_DN, 'cn=dummy', delold=0)
-    except ldap.CONSTRAINT_VIOLATION:
-        # yes it is expected
-        pass
-
-    # cleanup the stuff now
-    topology_st.standalone.delete_s(config.dn)
-    topology_st.standalone.delete_s(ACTIVE_USER_1_DN)
-    topology_st.standalone.delete_s(ACTIVE_USER_2_DN)
-
-
-def _active_stage_containers_add(topology_st, type_config='old', across_subtrees=False):
-    '''
-    Check uniqueness in several containers
-    Add an entry on a container with a given 'cn'
-    with across_subtrees=False check we CAN add an entry with the same 'cn' value on the other container
-    with across_subtrees=True check we CAN NOT add an entry with the same 'cn' value on the other container
-
-    '''
-    config = _build_config(topology_st, attr_name='cn', subtree_1=ACTIVE_DN, subtree_2=STAGE_DN,
-                           type_config=type_config, across_subtrees=False)
-
-    topology_st.standalone.add_s(config)
-    topology_st.standalone.restart(timeout=120)
-    topology_st.standalone.add_s(Entry((ACTIVE_USER_1_DN, {
-        'objectclass': "top person".split(),
-        'sn': ACTIVE_USER_1_CN,
-        'cn': ACTIVE_USER_1_CN})))
-    try:
-
-        # adding an entry on a separated contains with the same 'cn'
-        topology_st.standalone.add_s(Entry((STAGE_USER_1_DN, {
-            'objectclass': "top person".split(),
-            'sn': STAGE_USER_1_CN,
-            'cn': ACTIVE_USER_1_CN})))
-    except ldap.CONSTRAINT_VIOLATION:
-        assert across_subtrees
-
-    # cleanup the stuff now
-    topology_st.standalone.delete_s(config.dn)
-    topology_st.standalone.delete_s(ACTIVE_USER_1_DN)
-    topology_st.standalone.delete_s(STAGE_USER_1_DN)
-
-
-def _active_stage_containers_mod(topology_st, type_config='old', across_subtrees=False):
-    '''
-    Check uniqueness in a several containers
-    Add an entry on a container with a given 'cn', then check we CAN mod an entry with the same 'cn' value on the other container
-
-    '''
-    config = _build_config(topology_st, attr_name='cn', subtree_1=ACTIVE_DN, subtree_2=STAGE_DN,
-                           type_config=type_config, across_subtrees=False)
-
-    topology_st.standalone.add_s(config)
-    topology_st.standalone.restart(timeout=120)
-    # adding an entry on active with a different 'cn'
-    topology_st.standalone.add_s(Entry((ACTIVE_USER_1_DN, {
-        'objectclass': "top person".split(),
-        'sn': ACTIVE_USER_1_CN,
-        'cn': ACTIVE_USER_2_CN})))
-
-    # adding an entry on a stage with a different 'cn'
-    topology_st.standalone.add_s(Entry((STAGE_USER_1_DN, {
-        'objectclass': "top person".split(),
-        'sn': STAGE_USER_1_CN,
-        'cn': STAGE_USER_1_CN})))
-
-    try:
-
-        # modify add same value
-        topology_st.standalone.modify_s(STAGE_USER_1_DN, [(ldap.MOD_ADD, 'cn', [ensure_bytes(ACTIVE_USER_2_CN)])])
-    except ldap.CONSTRAINT_VIOLATION:
-        assert across_subtrees
-
-    topology_st.standalone.delete_s(STAGE_USER_1_DN)
-    topology_st.standalone.add_s(Entry((STAGE_USER_1_DN, {
-        'objectclass': "top person".split(),
-        'sn': STAGE_USER_1_CN,
-        'cn': STAGE_USER_2_CN})))
-    try:
-        # modify replace same value
-        topology_st.standalone.modify_s(STAGE_USER_1_DN,
-                                        [(ldap.MOD_REPLACE, 'cn', [ensure_bytes(STAGE_USER_2_CN), ensure_bytes(ACTIVE_USER_1_CN)])])
-    except ldap.CONSTRAINT_VIOLATION:
-        assert across_subtrees
-
-    # cleanup the stuff now
-    topology_st.standalone.delete_s(config.dn)
-    topology_st.standalone.delete_s(ACTIVE_USER_1_DN)
-    topology_st.standalone.delete_s(STAGE_USER_1_DN)
-
-
-def _active_stage_containers_modrdn(topology_st, type_config='old', across_subtrees=False):
-    '''
-    Check uniqueness in a several containers
-    Add and entry with a given 'cn', then check we CAN modrdn an entry with the same 'cn' value on the other container
-
-    '''
-
-    config = _build_config(topology_st, attr_name='cn', subtree_1=ACTIVE_DN, subtree_2=STAGE_DN,
-                           type_config=type_config, across_subtrees=False)
-
-    # enable the 'cn' uniqueness on Active and Stage
-    topology_st.standalone.add_s(config)
-    topology_st.standalone.restart(timeout=120)
-    topology_st.standalone.add_s(Entry((ACTIVE_USER_1_DN, {
-        'objectclass': "top person".split(),
-        'sn': ACTIVE_USER_1_CN,
-        'cn': [ACTIVE_USER_1_CN, 'dummy']})))
-
-    topology_st.standalone.add_s(Entry((STAGE_USER_1_DN, {
-        'objectclass': "top person".split(),
-        'sn': STAGE_USER_1_CN,
-        'cn': STAGE_USER_1_CN})))
-
-    try:
-
-        topology_st.standalone.rename_s(STAGE_USER_1_DN, 'cn=dummy', delold=0)
-
-        # check stage entry has 'cn=dummy'
-        stage_ent = topology_st.standalone.getEntry("cn=dummy,%s" % (STAGE_DN), ldap.SCOPE_BASE, "objectclass=*",
-                                                    ['cn'])
-        assert stage_ent.hasAttr('cn')
-        found = False
-        for value in stage_ent.getValues('cn'):
-            if ensure_str(value) == 'dummy':
-                found = True
-        assert found
-
-        # check active entry has 'cn=dummy'
-        active_ent = topology_st.standalone.getEntry(ACTIVE_USER_1_DN, ldap.SCOPE_BASE, "objectclass=*", ['cn'])
-        assert active_ent.hasAttr('cn')
-        found = False
-        for value in stage_ent.getValues('cn'):
-            if ensure_str(value) == 'dummy':
-                found = True
-        assert found
-
-        topology_st.standalone.delete_s("cn=dummy,%s" % (STAGE_DN))
-    except ldap.CONSTRAINT_VIOLATION:
-        assert across_subtrees
-        topology_st.standalone.delete_s(STAGE_USER_1_DN)
-
-    # cleanup the stuff now
-    topology_st.standalone.delete_s(config.dn)
-    topology_st.standalone.delete_s(ACTIVE_USER_1_DN)
-
-
-def _config_file(topology_st, action='save'):
-    dse_ldif = topology_st.standalone.confdir + '/dse.ldif'
-    sav_file = topology_st.standalone.confdir + '/dse.ldif.ticket47823'
-    if action == 'save':
-        shutil.copy(dse_ldif, sav_file)
-    else:
-        shutil.copy(sav_file, dse_ldif)
-    time.sleep(1)
-
-
-def _pattern_errorlog(file, log_pattern):
-    try:
-        _pattern_errorlog.last_pos += 1
-    except AttributeError:
-        _pattern_errorlog.last_pos = 0
-
-    found = None
-    log.debug("_pattern_errorlog: start at offset %d" % _pattern_errorlog.last_pos)
-    file.seek(_pattern_errorlog.last_pos)
-
-    # Use a while true iteration because 'for line in file: hit a
-    # python bug that break file.tell()
-    while True:
-        line = file.readline()
-        log.debug("_pattern_errorlog: [%d] %s" % (file.tell(), line))
-        found = log_pattern.search(line)
-        if ((line == '') or (found)):
-            break
-
-    log.debug("_pattern_errorlog: end at offset %d" % file.tell())
-    _pattern_errorlog.last_pos = file.tell()
-    return found
-
-
-def test_ticket47823_init(topology_st):
-    """
-
-    """
-
-    # Enabled the plugins
-    topology_st.standalone.plugins.enable(name=PLUGIN_ATTR_UNIQUENESS)
-    topology_st.standalone.restart(timeout=120)
-
-    topology_st.standalone.add_s(Entry((PROVISIONING_DN, {'objectclass': "top nscontainer".split(),
-                                                          'cn': PROVISIONING_CN})))
-    topology_st.standalone.add_s(Entry((ACTIVE_DN, {'objectclass': "top nscontainer".split(),
-                                                    'cn': ACTIVE_CN})))
-    topology_st.standalone.add_s(Entry((STAGE_DN, {'objectclass': "top nscontainer".split(),
-                                                   'cn': STAGE_CN})))
-    topology_st.standalone.add_s(Entry((DELETE_DN, {'objectclass': "top nscontainer".split(),
-                                                    'cn': DELETE_CN})))
-    topology_st.standalone.errorlog_file = open(topology_st.standalone.errlog, "r")
-
-    topology_st.standalone.stop(timeout=120)
-    time.sleep(1)
-    topology_st.standalone.start(timeout=120)
-    time.sleep(3)
-
-
-def test_ticket47823_one_container_add(topology_st):
-    '''
-    Check uniqueness in a single container
-    Add and entry with a given 'cn', then check we can not add an entry with the same 'cn' value
-
-    '''
-    _header(topology_st, "With former config (args), check attribute uniqueness with 'cn' (ADD) ")
-
-    _active_container_add(topology_st, type_config='old')
-
-    _header(topology_st, "With new config (args), check attribute uniqueness with 'cn' (ADD) ")
-
-    _active_container_add(topology_st, type_config='new')
-
-
-def test_ticket47823_one_container_mod(topology_st):
-    '''
-    Check uniqueness in a single container
-    Add and entry with a given 'cn', then check we can not modify an entry with the same 'cn' value
-
-    '''
-    _header(topology_st, "With former config (args), check attribute uniqueness with 'cn' (MOD)")
-
-    _active_container_mod(topology_st, type_config='old')
-
-    _header(topology_st, "With new config (args), check attribute uniqueness with 'cn' (MOD)")
-
-    _active_container_mod(topology_st, type_config='new')
-
-
-def test_ticket47823_one_container_modrdn(topology_st):
-    '''
-    Check uniqueness in a single container
-    Add and entry with a given 'cn', then check we can not modrdn an entry with the same 'cn' value
-
-    '''
-    _header(topology_st, "With former config (args), check attribute uniqueness with 'cn' (MODRDN)")
-
-    _active_container_modrdn(topology_st, type_config='old')
-
-    _header(topology_st, "With former config (args), check attribute uniqueness with 'cn' (MODRDN)")
-
-    _active_container_modrdn(topology_st, type_config='new')
-
-
-def test_ticket47823_multi_containers_add(topology_st):
-    '''
-    Check uniqueness in a several containers
-    Add and entry with a given 'cn', then check we can not add an entry with the same 'cn' value
-
-    '''
-    _header(topology_st, "With former config (args), check attribute uniqueness with 'cn' (ADD) ")
-
-    _active_stage_containers_add(topology_st, type_config='old', across_subtrees=False)
-
-    _header(topology_st, "With new config (args), check attribute uniqueness with 'cn' (ADD) ")
-
-    _active_stage_containers_add(topology_st, type_config='new', across_subtrees=False)
-
-
-def test_ticket47823_multi_containers_mod(topology_st):
-    '''
-    Check uniqueness in a several containers
-    Add an entry on a container with a given 'cn', then check we CAN mod an entry with the same 'cn' value on the other container
-
-    '''
-    _header(topology_st, "With former config (args), check attribute uniqueness with 'cn' (MOD) on separated container")
-
-    topology_st.standalone.log.info(
-        'Uniqueness not enforced: if same \'cn\' modified (add/replace) on separated containers')
-    _active_stage_containers_mod(topology_st, type_config='old', across_subtrees=False)
-
-    _header(topology_st, "With new config (args), check attribute uniqueness with 'cn' (MOD) on separated container")
-
-    topology_st.standalone.log.info(
-        'Uniqueness not enforced: if same \'cn\' modified (add/replace) on separated containers')
-    _active_stage_containers_mod(topology_st, type_config='new', across_subtrees=False)
-
-
-def test_ticket47823_multi_containers_modrdn(topology_st):
-    '''
-    Check uniqueness in a several containers
-    Add and entry with a given 'cn', then check we CAN modrdn an entry with the same 'cn' value on the other container
-
-    '''
-    _header(topology_st,
-            "With former config (args), check attribute uniqueness with 'cn' (MODRDN) on separated containers")
-
-    topology_st.standalone.log.info('Uniqueness not enforced: checks MODRDN entry is accepted on separated containers')
-    _active_stage_containers_modrdn(topology_st, type_config='old', across_subtrees=False)
-
-    topology_st.standalone.log.info('Uniqueness not enforced: checks MODRDN entry is accepted on separated containers')
-    _active_stage_containers_modrdn(topology_st, type_config='old')
-
-
-def test_ticket47823_across_multi_containers_add(topology_st):
-    '''
-    Check uniqueness across several containers, uniquely with the new configuration
-    Add and entry with a given 'cn', then check we can not add an entry with the same 'cn' value
-
-    '''
-    _header(topology_st, "With new config (args), check attribute uniqueness with 'cn' (ADD) across several containers")
-
-    _active_stage_containers_add(topology_st, type_config='old', across_subtrees=True)
-
-
-def test_ticket47823_across_multi_containers_mod(topology_st):
-    '''
-    Check uniqueness across several containers, uniquely with the new configuration
-    Add and entry with a given 'cn', then check we can not modifiy an entry with the same 'cn' value
-
-    '''
-    _header(topology_st, "With new config (args), check attribute uniqueness with 'cn' (MOD) across several containers")
-
-    _active_stage_containers_mod(topology_st, type_config='old', across_subtrees=True)
-
-
-def test_ticket47823_across_multi_containers_modrdn(topology_st):
-    '''
-    Check uniqueness across several containers, uniquely with the new configuration
-    Add and entry with a given 'cn', then check we can not modrdn an entry with the same 'cn' value
-
-    '''
-    _header(topology_st,
-            "With new config (args), check attribute uniqueness with 'cn' (MODRDN) across several containers")
-
-    _active_stage_containers_modrdn(topology_st, type_config='old', across_subtrees=True)
-
-
-def test_ticket47823_invalid_config_1(topology_st):
-    '''
-    Check that an invalid config is detected. No uniqueness enforced
-    Using old config: arg0 is missing
-    '''
-    _header(topology_st, "Invalid config (old): arg0 is missing")
-
-    _config_file(topology_st, action='save')
-
-    # create an invalid config without arg0
-    config = _build_config(topology_st, attr_name='cn', subtree_1=ACTIVE_DN, subtree_2=None, type_config='old',
-                           across_subtrees=False)
-
-    del config.data['nsslapd-pluginarg0']
-    # replace 'cn' uniqueness entry
-    try:
-        topology_st.standalone.delete_s(config.dn)
-
-    except ldap.NO_SUCH_OBJECT:
-        pass
-    topology_st.standalone.add_s(config)
-
-    topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-
-    # Check the server did not restart
-    topology_st.standalone.modify_s(DN_CONFIG, [(ldap.MOD_REPLACE, 'nsslapd-errorlog-level', b'65536')])
-    try:
-        topology_st.standalone.restart()
-        ent = topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)",
-                                              ALL_CONFIG_ATTRS)
-        if ent:
-            # be sure to restore a valid config before assert
-            _config_file(topology_st, action='restore')
-        assert not ent
-    except:
-        pass
-
-    # Check the expected error message
-    regex = re.compile("[U|u]nable to parse old style")
-    res = _pattern_errorlog(topology_st.standalone.errorlog_file, regex)
-    if not res:
-        # be sure to restore a valid config before assert
-        _config_file(topology_st, action='restore')
-    assert res
-
-    # Check we can restart the server
-    _config_file(topology_st, action='restore')
-    topology_st.standalone.start()
-    try:
-        topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-    except ldap.NO_SUCH_OBJECT:
-        pass
-
-
-def test_ticket47823_invalid_config_2(topology_st):
-    '''
-    Check that an invalid config is detected. No uniqueness enforced
-    Using old config: arg1 is missing
-    '''
-    _header(topology_st, "Invalid config (old): arg1 is missing")
-
-    _config_file(topology_st, action='save')
-
-    # create an invalid config without arg0
-    config = _build_config(topology_st, attr_name='cn', subtree_1=ACTIVE_DN, subtree_2=None, type_config='old',
-                           across_subtrees=False)
-
-    del config.data['nsslapd-pluginarg1']
-    # replace 'cn' uniqueness entry
-    try:
-        topology_st.standalone.delete_s(config.dn)
-
-    except ldap.NO_SUCH_OBJECT:
-        pass
-    topology_st.standalone.add_s(config)
-
-    topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-
-    # Check the server did not restart
-    try:
-        topology_st.standalone.restart()
-        ent = topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)",
-                                              ALL_CONFIG_ATTRS)
-        if ent:
-            # be sure to restore a valid config before assert
-            _config_file(topology_st, action='restore')
-        assert not ent
-    except:
-        pass
-
-    # Check the expected error message
-    regex = re.compile("No valid subtree is defined")
-    res = _pattern_errorlog(topology_st.standalone.errorlog_file, regex)
-    if not res:
-        # be sure to restore a valid config before assert
-        _config_file(topology_st, action='restore')
-    assert res
-
-    # Check we can restart the server
-    _config_file(topology_st, action='restore')
-    topology_st.standalone.start()
-    try:
-        topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-    except ldap.NO_SUCH_OBJECT:
-        pass
-
-
-def test_ticket47823_invalid_config_3(topology_st):
-    '''
-    Check that an invalid config is detected. No uniqueness enforced
-    Using old config: arg0 is missing
-    '''
-    _header(topology_st, "Invalid config (old): arg0 is missing but new config attrname exists")
-
-    _config_file(topology_st, action='save')
-
-    # create an invalid config without arg0
-    config = _build_config(topology_st, attr_name='cn', subtree_1=ACTIVE_DN, subtree_2=None, type_config='old',
-                           across_subtrees=False)
-
-    del config.data['nsslapd-pluginarg0']
-    config.data['uniqueness-attribute-name'] = 'cn'
-    # replace 'cn' uniqueness entry
-    try:
-        topology_st.standalone.delete_s(config.dn)
-
-    except ldap.NO_SUCH_OBJECT:
-        pass
-    topology_st.standalone.add_s(config)
-
-    topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-
-    # Check the server did not restart
-    topology_st.standalone.modify_s(DN_CONFIG, [(ldap.MOD_REPLACE, 'nsslapd-errorlog-level', b'65536')])
-    try:
-        topology_st.standalone.restart()
-        ent = topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)",
-                                              ALL_CONFIG_ATTRS)
-        if ent:
-            # be sure to restore a valid config before assert
-            _config_file(topology_st, action='restore')
-        assert not ent
-    except:
-        pass
-
-    # Check the expected error message
-    regex = re.compile("[U|u]nable to parse old style")
-    res = _pattern_errorlog(topology_st.standalone.errorlog_file, regex)
-    if not res:
-        # be sure to restore a valid config before assert
-        _config_file(topology_st, action='restore')
-    assert res
-
-    # Check we can restart the server
-    _config_file(topology_st, action='restore')
-    topology_st.standalone.start()
-    try:
-        topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-    except ldap.NO_SUCH_OBJECT:
-        pass
-
-
-def test_ticket47823_invalid_config_4(topology_st):
-    '''
-    Check that an invalid config is detected. No uniqueness enforced
-    Using old config: arg1 is missing
-    '''
-    _header(topology_st, "Invalid config (old): arg1 is missing but new config exist")
-
-    _config_file(topology_st, action='save')
-
-    # create an invalid config without arg0
-    config = _build_config(topology_st, attr_name='cn', subtree_1=ACTIVE_DN, subtree_2=None, type_config='old',
-                           across_subtrees=False)
-
-    del config.data['nsslapd-pluginarg1']
-    config.data['uniqueness-subtrees'] = ACTIVE_DN
-    # replace 'cn' uniqueness entry
-    try:
-        topology_st.standalone.delete_s(config.dn)
-
-    except ldap.NO_SUCH_OBJECT:
-        pass
-    topology_st.standalone.add_s(config)
-
-    topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-
-    # Check the server did not restart
-    try:
-        topology_st.standalone.restart()
-        ent = topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)",
-                                              ALL_CONFIG_ATTRS)
-        if ent:
-            # be sure to restore a valid config before assert
-            _config_file(topology_st, action='restore')
-        assert not ent
-    except:
-        pass
-
-    # Check the expected error message
-    regex = re.compile("No valid subtree is defined")
-    res = _pattern_errorlog(topology_st.standalone.errorlog_file, regex)
-    if not res:
-        # be sure to restore a valid config before assert
-        _config_file(topology_st, action='restore')
-    assert res
-
-    # Check we can restart the server
-    _config_file(topology_st, action='restore')
-    topology_st.standalone.start()
-    try:
-        topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-    except ldap.NO_SUCH_OBJECT:
-        pass
-
-
-def test_ticket47823_invalid_config_5(topology_st):
-    '''
-    Check that an invalid config is detected. No uniqueness enforced
-    Using new config: uniqueness-attribute-name is missing
-    '''
-    _header(topology_st, "Invalid config (new): uniqueness-attribute-name is missing")
-
-    _config_file(topology_st, action='save')
-
-    # create an invalid config without arg0
-    config = _build_config(topology_st, attr_name='cn', subtree_1=ACTIVE_DN, subtree_2=None, type_config='new',
-                           across_subtrees=False)
-
-    del config.data['uniqueness-attribute-name']
-    # replace 'cn' uniqueness entry
-    try:
-        topology_st.standalone.delete_s(config.dn)
-
-    except ldap.NO_SUCH_OBJECT:
-        pass
-    topology_st.standalone.add_s(config)
-
-    topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-
-    # Check the server did not restart
-    try:
-        topology_st.standalone.restart()
-        ent = topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)",
-                                              ALL_CONFIG_ATTRS)
-        if ent:
-            # be sure to restore a valid config before assert
-            _config_file(topology_st, action='restore')
-        assert not ent
-    except:
-        pass
-
-    # Check the expected error message
-    regex = re.compile("[A|a]ttribute name not defined")
-    res = _pattern_errorlog(topology_st.standalone.errorlog_file, regex)
-    if not res:
-        # be sure to restore a valid config before assert
-        _config_file(topology_st, action='restore')
-    assert res
-
-    # Check we can restart the server
-    _config_file(topology_st, action='restore')
-    topology_st.standalone.start()
-    try:
-        topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-    except ldap.NO_SUCH_OBJECT:
-        pass
-
-
-def test_ticket47823_invalid_config_6(topology_st):
-    '''
-    Check that an invalid config is detected. No uniqueness enforced
-    Using new config: uniqueness-subtrees is missing
-    '''
-    _header(topology_st, "Invalid config (new): uniqueness-subtrees is missing")
-
-    _config_file(topology_st, action='save')
-
-    # create an invalid config without arg0
-    config = _build_config(topology_st, attr_name='cn', subtree_1=ACTIVE_DN, subtree_2=None, type_config='new',
-                           across_subtrees=False)
-
-    del config.data['uniqueness-subtrees']
-    # replace 'cn' uniqueness entry
-    try:
-        topology_st.standalone.delete_s(config.dn)
-
-    except ldap.NO_SUCH_OBJECT:
-        pass
-    topology_st.standalone.add_s(config)
-
-    topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-
-    # Check the server did not restart
-    try:
-        topology_st.standalone.restart()
-        ent = topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)",
-                                              ALL_CONFIG_ATTRS)
-        if ent:
-            # be sure to restore a valid config before assert
-            _config_file(topology_st, action='restore')
-        assert not ent
-    except:
-        pass
-
-    # Check the expected error message
-    regex = re.compile("[O|o]bjectclass for subtree entries is not defined")
-    res = _pattern_errorlog(topology_st.standalone.errorlog_file, regex)
-    if not res:
-        # be sure to restore a valid config before assert
-        _config_file(topology_st, action='restore')
-    assert res
-
-    # Check we can restart the server
-    _config_file(topology_st, action='restore')
-    topology_st.standalone.start()
-    try:
-        topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-    except ldap.NO_SUCH_OBJECT:
-        pass
-
-
-def test_ticket47823_invalid_config_7(topology_st):
-    '''
-    Check that an invalid config is detected. No uniqueness enforced
-    Using new config: uniqueness-subtrees is missing
-    '''
-    _header(topology_st, "Invalid config (new): uniqueness-subtrees are invalid")
-
-    _config_file(topology_st, action='save')
-
-    # create an invalid config without arg0
-    config = _build_config(topology_st, attr_name='cn', subtree_1="this_is dummy DN", subtree_2="an other=dummy DN",
-                           type_config='new', across_subtrees=False)
-
-    topology_st.standalone.modify_s(DN_CONFIG, [(ldap.MOD_REPLACE, 'nsslapd-errorlog-level', b'65536')])
-    # replace 'cn' uniqueness entry
-    try:
-        topology_st.standalone.delete_s(config.dn)
-
-    except ldap.NO_SUCH_OBJECT:
-        pass
-    topology_st.standalone.add_s(config)
-
-    topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-
-    # Check the server did not restart
-    try:
-        topology_st.standalone.restart()
-        ent = topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)",
-                                              ALL_CONFIG_ATTRS)
-        if ent:
-            # be sure to restore a valid config before assert
-            _config_file(topology_st, action='restore')
-        assert not ent
-    except:
-        pass
-
-    # Check the expected error message
-    regex = re.compile("No valid subtree is defined")
-    res = _pattern_errorlog(topology_st.standalone.errorlog_file, regex)
-    if not res:
-        # be sure to restore a valid config before assert
-        _config_file(topology_st, action='restore')
-    assert res
-
-    # Check we can restart the server
-    _config_file(topology_st, action='restore')
-    topology_st.standalone.start()
-    try:
-        topology_st.standalone.getEntry(config.dn, ldap.SCOPE_BASE, "(objectclass=nsSlapdPlugin)", ALL_CONFIG_ATTRS)
-    except ldap.NO_SUCH_OBJECT:
-        pass
-
-
-if __name__ == '__main__':
-    # Run isolated
-    # -s for DEBUG mode
-    CURRENT_FILE = os.path.realpath(__file__)
-    pytest.main("-s %s" % CURRENT_FILE)
-- 
2.52.0

